### PR TITLE
Mach cleanup 1

### PR DESCRIPTION
**Pull request by brentlu**
It's an effort to reduce the machine numbers in enumeration table. Now all boards with same headphone codec could share one entry in the enumeration table and sof driver will detect the type of headphone and amplifier then compose the topology file name at run time instead of hard-coding in the enumeration table and creating entries for each amplifier.

### CODE REVIEW

**Review by bardliao — DISMISSED** [review]

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: This looks like a circular dependency to me. It's problematic if the PCI parent module depends on the board level stuff.

**Review by plbossart — COMMENTED** [review]
Not a big fan on the new module dependency @brentlu @bardliao, see below.

**Comment by plbossart** [comment]
Also the nocodec build is broken

ERROR: modpost: "sof_ssp_get_codec_tplg_suffix" [sound/soc/sof/intel/snd-sof-intel-hda-generic.ko] undefined!
ERROR: modpost: "sof_ssp_get_amp_tplg_suffix" [sound/soc/sof/intel/snd-sof-intel-hda-generic.ko] undefined!

File: sound/soc/sof/intel/hda.c
Comment (brentlu) [reply sound/soc/sof/intel/hda.c]: How about moving the code of ssp-common to SND_SOC_ACPI_INTEL_MATCH module? They are similar in many ways. Both modules handle codec/amp detection via ACPI. Both are designed for machine driver enumeration.

Comment (plbossart) [reply sound/soc/sof/intel/hda.c]: That sounds ok at a high-level, we just need to make sure the dependencies are well managed and that this wouldn't break CI with the sof-test/tools/kmod scripts doing things that no longer work.

Comment (brentlu) [reply sound/soc/sof/intel/hda.c:1464]: Another PR #4852 is submitted

**Comment by brentlu** [comment]
The latest patches move code of ssp-common to snd_soc_acpi_intel_match to avoid creating new dependency between ssp-common and snd_sof_intel_hda_common.

Before merge the two modules:
snd_soc_acpi_intel_match    61440  2 snd_sof_intel_hda_common,snd_sof_pci_intel_tgl
snd_soc_intel_sof_ssp_common    16384  2 snd_soc_intel_sof_board_helpers,snd_sof_intel_hda_common

After:
snd_soc_acpi_intel_match    61440  3 snd_soc_intel_sof_board_helpers,snd_sof_intel_hda_common,snd_sof_pci_intel_tgl

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [inline sound/soc/sof/intel/hda.c:1412]: Humm, I am a bit worried that we are adding two more branches to change the topology filename. In the end there will be 3 suffixes, are they all independent?

I think the initial one  SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER can only be used for the codec, it's a single-SSP configuration.

Comment (brentlu) [reply sound/soc/sof/intel/hda.c:1412]: Yes, they are independent and currently SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER flag is used in es8336 driver enumeration only.

Other machine drivers (mostly Chromebooks) could use 'tplg_filename' module parameter to overwrite the topology file name if they need to use something like '-sspX' instead of using SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER flag.

Comment (plbossart) [reply sound/soc/sof/intel/hda.c:1412]: > Yes, they are independent and currently SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER flag is used in es8336 driver enumeration only.

The reply above tells me all flags cannot be used concurrently, so they are not independent flags with orthogonal logic.

Comment (brentlu) [reply sound/soc/sof/intel/hda.c:1412]: Sorry for my misconception. I thought the output suffix string of a flag does not depend on the code branch of other flags so they are independent.

Do you foresee a situation that we have to use the three flags concurrently?

Comment (plbossart) [reply sound/soc/sof/intel/hda.c:1412]: No, like you said the SSP_NUMBER quirks was only meant to be used for the ES8336. Just adding a test to check that SSP_NUMBER is ignored if one of the others is set should work and avoid broken configurations leading to 'no topology file found'.

The point is that this needs to be user-proof with defensive programming. Some folks will set the quirk as -1 just because they can...

Comment (brentlu) [reply sound/soc/sof/intel/hda.c:1412]: Ok. Let me test following code segment and upload for review tomorrow.

```
diff --git a/sound/soc/sof/intel/hda.c b/sound/soc/sof/intel/hda.c
index 67a4d50cb1d7..a490f15d46be 100644
--- a/sound/soc/sof/intel/hda.c
+++ b/sound/soc/sof/intel/hda.c
@@ -1261,6 +1261,27 @@ void hda_set_mach_params(struct snd_soc_acpi_mach *mach,
        mach_params->dai_drivers = desc->ops->drv;
 }

+static int check_tplg_quirk_mask(struct snd_soc_acpi_mach *mach)
+{
+       u32 dmic_ssp_quirk;
+       u32 codec_amp_name_quirk;
+
+       /*
+        * In current implementation dmic and ssp quirks are designed for es8336
+        * machine driver and could not be mixed with codec name and amp name
+        * quirks.
+        */
+       dmic_ssp_quirk = mach->tplg_quirk_mask &
+                        (SND_SOC_ACPI_TPLG_INTEL_DMIC_NUMBER | SND_SOC_ACPI_TPLG_INTEL_SSP_NUMBER);
+       codec_amp_name_quirk = mach->tplg_quirk_mask &
+                        (SND_SOC_ACPI_TPLG_INTEL_AMP_NAME | SND_SOC_ACPI_TPLG_INTEL_CODEC_NAME);
+
+       if (dmic_ssp_quirk && codec_amp_name_quirk)
+               return -EINVAL;
+
+       return 0;
+}
+
 struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
 {
        u32 interface_mask = hda_get_interface_mask(sdev);
@@ -1288,6 +1309,17 @@ struct snd_soc_acpi_mach *hda_machine_select(struct snd_sof_dev *sdev)
                        tplg_fixup = true;
                }

+               /*
+                * Checking quirk mask integrity; some quirk flags could not be
+                * set concurrently.
+                */
+               if (tplg_fixup &&
+                   check_tplg_quirk_mask(mach)) {
+                       dev_err(sdev->dev, "Invalid tplg quirk mask 0x%x\n",
+                               mach->tplg_quirk_mask);
+                       return NULL;
+               }
+
                /* report to machine driver if any DMICs are found */
                mach->mach_params.dmic_num = check_dmic_num(sdev);
```

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: let's drop the 'error' from new dev_error messages.

File: sound/soc/sof/intel/hda.c
Comment (plbossart) [inline sound/soc/sof/intel/hda.c]: is this needed? I thought we already logged the topology file name, I see it in all CI results.

**Review by plbossart — COMMENTED** [review]
two minor points below, LGTM otherwise.

File: sound/soc/sof/intel/hda.c
Comment (brentlu) [reply sound/soc/sof/intel/hda.c]: done

File: sound/soc/sof/intel/hda.c
Comment (brentlu) [reply sound/soc/sof/intel/hda.c]: I used this string for testing. Removed.

**Review by plbossart — APPROVED** [review]
@bardliao I'll let you review/merge if ok. I'd like to send this upstream next week with the large machine driver changes.

**Review by bardliao — APPROVED** [review]

**Pull request merged**
